comp/metadata/runner/runner.go

```diff
@@ -0,0 +1,149 @@
+
+package runner
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/comp/core/config"
+	"github.com/DataDog/datadog-agent/comp/core/log"
+	"github.com/DataDog/datadog-agent/pkg/serializer"
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"go.uber.org/fx"
+)
+
+type MetadataProvider func(context.Context, serializer.MetricSerializer) time.Duration
+
+type runner struct {
+	log    log.Component
+	config config.Component
+
+	// providers are the metada providers to run. They're Optional because some of them can be disabled through the
+	// configuration
+	providers []util.Optional[MetadataProvider] `group:"metadata_provider"`
+
+	wg       sync.WaitGroup
+	stopChan chan struct{}
+
+	serializer serializer.MetricSerializer
+}
+
+type dependencies struct {
+	fx.In
+
+	Log        log.Component
+	Config     config.Component
+	Serializer serializer.MetricSerializer
+
+	Providers []util.Optional[MetadataProvider] `group:"metadata_provider"`
+}
+
+// Provider represents the callback from a metada provider. This is returned by 'NewProvider' helper.
+type Provider struct {
+	fx.Out
+
+	Callback util.Optional[MetadataProvider] `group:"metadata_provider"`
+}
+
+// NewEmptyProvider returns a empty provider which is not going to register anything. This is useful for providers that
+// can be enabled/disabled through configuration.
+func NewEmptyProvider() Provider {
+	return Provider{
+		Callback: util.NewNoneOptional[MetadataProvider](),
+	}
+}
+
+// NewProvider registers a new metadata provider by adding a callback to the runner.
+func NewProvider(callback MetadataProvider) Provider {
+	return Provider{
+		Callback: util.NewOptional[MetadataProvider](callback),
+	}
+}
+
+// createRunner instantiates a runner object
+func createRunner(deps dependencies) *runner {
+	return &runner{
+		log:        deps.Log,
+		config:     deps.Config,
+		providers:  deps.Providers,
+		serializer: deps.Serializer,
+		stopChan:   make(chan struct{}),
+	}
+}
+
+func newRunner(lc fx.Lifecycle, deps dependencies) Component {
+	r := createRunner(deps)
+
+	// We rely on FX to start and stop the metadata runner
+	lc.Append(fx.Hook{
+		OnStart: func(ctx context.Context) error {
+			return r.start()
+		},
+		OnStop: func(ctx context.Context) error {
+			return r.stop()
+		},
+	})
+	return r
+}
+
+// handleProvider runs a provider at regular interval until the runner is stopped
+func (r *runner) handleProvider(p MetadataProvider) {
+	r.log.Debugf("Starting runner for MetadataProvider %#v", p)
+	r.wg.Add(1)
+
+	intervalChan := make(chan time.Duration)
+	var interval time.Duration
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer func() {
+		cancel()
+		r.log.Debugf("stopping runner for MetadataProvider %#v", p)
+		r.wg.Done()
+	}()
+
+	for {
+		go func(intervalChan chan time.Duration) {
+			intervalChan <- p(ctx, r.serializer)
+		}(intervalChan)
+
+		select {
+		case interval = <-intervalChan:
+		case <-r.stopChan:
+			cancel()
```

Comment:
Note: This is strictly not needed as `cancel` is called [here](https://github.com/DataDog/datadog-agent/pull/18099/files#diff-effd0202d7ac04db2a5ed611246a121d1d512c137dfb2288c9e1b250427d7288R104)